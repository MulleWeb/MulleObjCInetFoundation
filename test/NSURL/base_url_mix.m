//
//  main.m
//  NSURLTest
//
//  Created by Nat! on 24.02.20.
//  Copyright Â© 2020 Nat!. All rights reserved.
//
#ifdef __MULLE_OBJC__
# import <MulleObjCInetFoundation/MulleObjCInetFoundation.h>
# include <mulle-stacktrace/mulle-stacktrace.h>
#else
# import <Foundation/Foundation.h>
#endif

static void   run_test( NSURL *baseURL, NSString *string)
{
   NSURL      *URL;
   NSString   *result;

   URL = [NSURL URLWithString:string
                relativeToURL:baseURL];
   result = [URL absoluteString];
   if( ! URL && ! [NSURL URLWithString:string])
      result = @"relative is invalid";

   printf( "%s (base)\n + %s (relative)\n-> %s\n\n", [[baseURL absoluteString] UTF8String],
                                   [string UTF8String],
                                   [result UTF8String]);
}


static void   test( NSString *baseURLString)
{
   NSString   *string;
   NSURL      *baseURL;

   baseURL = [NSURL URLWithString:baseURLString];
   if( ! baseURL)
   {
      printf( "baseURL %s is invalid\n", [baseURLString UTF8String]);
      return;
   }

   string = @"http://"
             "user:password"
             "@host:32765"
             "/path"
             ";parameter"
             "?query"
             "#fragment";
   run_test( baseURL, string);

   string  = @"/rel_path/foo.html"
              "#fragment";
   run_test( baseURL, string);

   string = @"rel_path/bar.html"
             ";parameter"
             "#fragment";
   run_test( baseURL, string);

   string = @";parameter";
   run_test( baseURL, string);
}

/*
  Since we don't have a HTTP parser without MulleObjCHTTPFoundation the baseURL
  results look a little counterintuitive sometimes, but I think it is OK
 */
int   main(int argc, const char * argv[])
{
   NSString   *string;

#ifdef __MULLE_OBJC__
   struct _mulle_objc_universe   *universe;

   universe = mulle_objc_global_get_universe( __MULLE_OBJC_UNIVERSEID__);
   universe->debug.count_stackdepth = mulle_stacktrace_count_frames;
#endif

   string = @"http://"
            "base-user:base-password"
            "@base-host:80"
            "/base-path/"
            ";base-parameter"
            "?base-query"
            "#base-fragment";

   test( string);

   string = @"http://"
             "base-host:80"
             "/base-path"
             "?base-query"
             "#base-fragment";
   test( string);

   return 0;
}


