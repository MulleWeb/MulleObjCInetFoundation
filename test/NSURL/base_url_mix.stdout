http://base-user:base-password@base-host:80/base-path/;base-parameter?base-query#base-fragment (base)
 + http://user:password@host:32765/path;parameter?query#fragment (relative)
-> http://user:password@host:32765/path;parameter?query#fragment

http://base-user:base-password@base-host:80/base-path/;base-parameter?base-query#base-fragment (base)
 + /rel_path/foo.html#fragment (relative)
-> http:/rel_path/foo.html#fragment

http://base-user:base-password@base-host:80/base-path/;base-parameter?base-query#base-fragment (base)
 + rel_path/bar.html;parameter#fragment (relative)
-> http://base-user:base-password@base-host:80/base-path/rel_path/bar.html;parameter#fragment

http://base-user:base-password@base-host:80/base-path/;base-parameter?base-query#base-fragment (base)
 + ;parameter (relative)
-> http://base-user:base-password@base-host:80/base-path/;parameter

http://base-host:80/base-path?base-query#base-fragment (base)
 + http://user:password@host:32765/path;parameter?query#fragment (relative)
-> http://user:password@host:32765/path;parameter?query#fragment

http://base-host:80/base-path?base-query#base-fragment (base)
 + /rel_path/foo.html#fragment (relative)
-> http:/rel_path/foo.html#fragment

http://base-host:80/base-path?base-query#base-fragment (base)
 + rel_path/bar.html;parameter#fragment (relative)
-> http://base-host:80/rel_path/bar.html;parameter#fragment

http://base-host:80/base-path?base-query#base-fragment (base)
 + ;parameter (relative)
-> http://base-host:80/base-path;parameter

